### PR TITLE
Cumulative weight based entry point selector

### LINKED ISSUE #50: Move entry point to more recent past (heuristic)

### PR DESCRIPTION
**Pull request by alongalky**
Fixes #50

### CODE REVIEW

File: src/main/java/com/iota/iri/service/tipselection/impl/EntryPointSelectorCumulativeWeightThreshold.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+import org.apache.commons.lang3.NotImplementedException;
+
+/**
+ * Implementation of <tt>EntryPointSelector</tt> that backtracks via trunkTransaction
+ * until reaching a minimum Cumulative Weight or the genesis.
+ */
+public class EntryPointSelectorCumulativeWeightThreshold implements EntryPointSelector {
+    public final Tangle tangle;
+    private final CumulativeWeightCalculator cumulativeWeightCalculator;
```
Comment (alon-e) [inline src/main/java/com/iota/iri/service/tipselection/impl/EntryPointSelectorCumulativeWeightThreshold.java:19]: why not use the interface `RatingCalculator`?

File: src/main/java/com/iota/iri/service/tipselection/impl/EntryPointSelectorCumulativeWeightThreshold.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    @Override
+    public Hash getEntryPoint() throws Exception {
+        return backtrack(tipsViewModel.getRandomSolidTipHash(), threshold);
+    }
+
+    public Hash backtrack(Hash tip, int threshold) throws Exception {
+        Hash currentHash = tip;
+        Integer currentWeight = 0;
+
+        // Stop when reaching the genesis or the cumulative weight threshold
```
Comment (alon-e) [inline src/main/java/com/iota/iri/service/tipselection/impl/EntryPointSelectorCumulativeWeightThreshold.java]: this doc line is confusing - it's inverted from what is actually written in the code.

File: src/main/java/com/iota/iri/service/tipselection/impl/EntryPointSelectorCumulativeWeightThreshold.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        while (currentWeight < threshold && !Objects.equals(currentHash, Hash.NULL_HASH)) {
+            currentHash = TransactionViewModel.fromHash(tangle, currentHash).getTrunkTransactionHash();
+            currentWeight = cumulativeWeightCalculator.calculate(currentHash).get(currentHash);
```
Comment (alon-e) [inline src/main/java/com/iota/iri/service/tipselection/impl/EntryPointSelectorCumulativeWeightThreshold.java]: I think this method should use the simplified version of CW calculation - via b/dfs. the method used assumes you would use all intermediate values.

my proposal is to add a method (to the `RatingCalculator` interface)  that returns the weight of a transaction - and runs the simplified version.

File: src/test/java/com/iota/iri/service/tipselection/impl/EntryPointSelectorCumulativeWeightThresholdTest.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        TransactionViewModel transaction;
+        transaction = new TransactionViewModel(getRandomTransactionTrits(), getRandomTransactionHash());
+        transaction.store(tangle);
+        int threshold = 50;
+        Mockito.when(tipsViewModel.getRandomSolidTipHash()).thenReturn(transaction.getBundleHash());
+        
+        EntryPointSelector entryPointSelector = new EntryPointSelectorCumulativeWeightThreshold(tangle, tipsViewModel, threshold);
+        Hash entryPoint = entryPointSelector.getEntryPoint();
+
+        Assert.assertEquals(Hash.NULL_HASH, entryPoint);
```
Comment (alon-e) [inline src/test/java/com/iota/iri/service/tipselection/impl/EntryPointSelectorCumulativeWeightThresholdTest.java:59]: I don't get how this returns Genesis -- your tip does not point to Genesis, I think it had to do w/ a bug that when a `TxVM.fromHash(UNKOWN_TX)` returns all-9

**Review by alon-e — CHANGES_REQUESTED** [review]

File: src/test/java/com/iota/iri/service/tipselection/impl/EntryPointSelectorCumulativeWeightThresholdTest.java (hunk 4 above)
Comment (alongalky) [reply src/test/java/com/iota/iri/service/tipselection/impl/EntryPointSelectorCumulativeWeightThresholdTest.java:59]: I'll correct the test

File: src/main/java/com/iota/iri/service/tipselection/impl/EntryPointSelectorCumulativeWeightThreshold.java (hunk 2 above)
Comment (alongalky) [reply src/main/java/com/iota/iri/service/tipselection/impl/EntryPointSelectorCumulativeWeightThreshold.java]: Changed to `Backtrack as long as the genesis hasn't been reached and the tresholed has not been crossed`

File: src/main/java/com/iota/iri/service/tipselection/impl/EntryPointSelectorCumulativeWeightThreshold.java (hunk 1 above)
Comment (alongalky) [reply src/main/java/com/iota/iri/service/tipselection/impl/EntryPointSelectorCumulativeWeightThreshold.java:19]: I don't want a generic rating (with LM / stars / whatever), I want CW specifically

**Review by alon-e — APPROVED** [review]

**Pull request merged**